### PR TITLE
feat(trainer): Add GPU passthrough support for container backend

### LINKED ISSUE #159: Enable GPU Support in the Kubeflow SDK Container Backend
### What you would like to be added?

This issue proposes adding support for GPU-accelerated workloads in the Kubeflow SDK’s container backend, across both container runtimes: Docker and Podman.

### Podman Ecosystem
There are several Podman-related projects that may help provide GPU capabilities or improved virtualization features:
-**libkrun**
[ https://github.com/containers/libkrun](https://github.com/containers/libkrun)
-**krunkit**
[ https://github.com/containers/krunkit](https://github.com/containers/krunkit)
-**Podman Desktop GPU documentation**
[ https://podman-desktop.io/docs/podman/gpu](https://podman-desktop.io/docs/podman/gpu)

### Docker / NVIDIA Support
Docker already has established ways to expose GPUs:
--gpus flag via the NVIDIA Container Toolkit

Platform-specific support (Linux native, WSL2 GPU on Windows, etc.)

The SDK doesn’t currently expose an abstraction that allows using these features in a unified way.

### PyTorch MPS (macOS-specific context)
Ideally, PyTorch’s MPS backend could work inside a containerized workflow:
- https://developer.apple.com/metal/pytorch/
- https://docs.pytorch.org/docs/stable/notes/mps.html

However, current understanding is that this cannot work unless MPS becomes available in Linux containers

### Why is this needed?

Many Kubeflow users want to run GPU-enabled ML workloads locally or in lightweight environments using Docker or Podman. Today, the SDK’s container backend largely treats all runtimes the same, and does not expose or leverage runtime-specific GPU support options.
Supporting GPU passthrough (or other forms of GPU access) would significantly improve the development experience for users building workflows that rely on PyTorch, TensorFlow, JAX, or similar frameworks.

### Love this feature?

Give it a 👍 We prioritize the features with most 👍

### PR DESCRIPTION
**Pull request by muzzlol**
**What this PR does / why we need it**:

Fixes #159

## Changes

- BaseContainerClientAdapter: Added `gpu_count` parameter to `create_and_start_container`
- DockerClientAdapter: Configures GPU access via `device_requests`
- PodmanClientAdapter: Configures GPU access via `devices` parameter
- ContainerBackend: Extracts GPU count from `resources_per_node`, handles macOS warning
- Tests: Added 5 test cases covering GPU passthrough scenarios

**Checklist:**

- [x] All  pass (make test-python)
- [x] Linting passes (make verify)
- [x] Documentation added (docstrings for new parameters)
- [x] Tests cover new functionality
- [x] No breaking changes to public APIs (backward compatible - gpu_count is optional)
- [x] Tested locally with both runtimes (docker, podman) on NVIDIA GPU (Turning arch)

Things to mention in docs relating to gpu compatability in containers: 
  - Docker: Install NVIDIA Container Toolkit (nvidia-ctk runtime configure --runtime=docker)
  - Podman: CDI  (configured by default with NVIDIA Container Toolkit v1.12.0+)
  - Linux and Windows via wsl2 only: macOS users will see a warning that GPU passthrough is unavailable
  - RHEL/Fedora users with SELinux enforcing may need to run with --security-opt=label=disable if they encounter NVML: Insufficient Permissions errors. according [to](RHEL/Fedora users with SELinux enforcing may need to run with --security-opt=label=disable if they encounter NVML: Insufficient Permissions errors.)

**Relevant sources:**
https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/index.html
https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html
https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/cdi-support.html
https://podman-desktop.io/docs/podman/gpu

### CODE REVIEW

**Comment by muzzlol** [comment]
@Fiona-Waters 
I couldn’t find any existing SDK docs, so I included the relevant sources and references for maintainers to use or place as needed at the end of the PR desc.

File: kubeflow/trainer/backends/container/backend.py [hunk 1]
```diff
@@ -284,10 +285,22 @@ def train(
...
+
+                    # Check for macOS - GPU passthrough is not supported
+                    if platform.system() == "Darwin":
+                        logger.warning(
+                            "GPU passthrough is not supported on macOS. "
+                            "Containers will run without GPU access. "
+                            "To use GPUs, run on a Linux machine with NVIDIA drivers "
+                            "and the NVIDIA Container Toolkit installed."
+                        )
+                        gpu_count = None  # Don't attempt GPU passthrough on macOS
```
Comment (Fiona-Waters) [inline kubeflow/trainer/backends/container/backend.py:303]: @muzzlol Did you look into PyTorch MPS for Mac OS?

**Review by Fiona-Waters — COMMENTED** [review]
Thank you for this @muzzlol ! I have left one question, and also we definitely need to add docs. There is an issue related to where docs should go [here](https://github.com/kubeflow/sdk/issues/50). @kramaranya in the meantime should we create a follow on issue for these docs?

**Comment by kramaranya** [comment]
> Thank you for this @muzzlol ! I have left one question, and also we definitely need to add docs. There is an issue related to where docs should go [here](https://github.com/kubeflow/sdk/issues/50). @kramaranya in the meantime should we create a follow on issue for these docs?

@muzzlol you can open PRs to update the [website](https://www.kubeflow.org/docs/components/trainer/user-guides/local-execution-mode/) and trainer [examples](https://github.com/kubeflow/trainer/tree/master/examples/local)

File: kubeflow/trainer/backends/container/backend.py (hunk 1 above)
Comment (muzzlol) [reply kubeflow/trainer/backends/container/backend.py:303]: Yep. GPU passthrough for containers is still not supported on macOS (https://github.com/apple/container/discussions/62) so PyTorch MPS is a no-go unfortunately. 
There is a discussion around utilizing PyTorch's Vulkan backend [here](https://github.com/apple/container/discussions/62#discussioncomment-15343816), but afaik that backend supports mobile/inference-only, not training.

Comment (andreyvelich) [reply kubeflow/trainer/backends/container/backend.py:303]: Looking at the comment the ML training is also supported, isn't? 
Have you tried to build PyTorch with Vulkan backend as suggested here: https://docs.pytorch.org/tutorials/unstable/vulkan_workflow.html#building-pytorch-with-vulkan-backend ?

Comment (muzzlol) [reply kubeflow/trainer/backends/container/backend.py:303]: > Looking at the comment the ML training is also supported, isn't?

The comment seems to be a bit misleading. [check](https://github.com/wisejnrs/wisejnrs-carbon-runtime/blob/main/COMPLETE-MACOS-GUIDE.md#-complete-feature-comparison)



> Have you tried to build PyTorch with Vulkan backend as suggested here: 
https://docs.pytorch.org/tutorials/unstable/vulkan_workflow.html#building-pytorch-with-vulkan-backend ?

No I have not. I currently do not have access to my mac machine, but I'm happy to try it out tommorow if required.

Comment (andreyvelich) [reply kubeflow/trainer/backends/container/backend.py:303]: I think, it would be nice to try this out, and see whether simple PyTorch training examples (e.g. MNIST) will be working.

Comment (muzzlol) [reply kubeflow/trainer/backends/container/backend.py:303]: > I think, it would be nice to try this out, and see whether simple PyTorch training examples (e.g. MNIST) will be working.

I tested and the vulkan backend does not have autograd support. Shall I proceed with https://github.com/kubeflow/sdk/pull/219#issuecomment-3711085734 ?
